src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -2221,114 +2202,100 @@ impl Coordinator {
         }
     }
 
+    /// Determines the query timestamp and acquires read holds on dependent sources
+    /// if necessary.
+    fn sequence_peek_timestamp(
+        &mut self,
+        session: &mut Session,
+        when: &QueryWhen,
+        cluster_id: ClusterId,
+        timeline_context: TimelineContext,
+        id_bundle: &CollectionIdBundle,
+        source_ids: &BTreeSet<GlobalId>,
+        real_time_recency_ts: Option<Timestamp>,
+    ) -> Result<TimestampDetermination<Timestamp>, AdapterError> {
+        let determination = self.determine_timestamp(
+            session,
+            id_bundle,
+            when,
+            cluster_id,
+            timeline_context.clone(),
+            real_time_recency_ts,
+        )?;
+
+        // We only track the peeks in the session if the query doesn't use AS
+        // OF or we're inside an explicit transaction. The latter case is
+        // necessary to support PG's `BEGIN` semantics, whose behavior can
+        // depend on whether or not reads have occurred in the txn.
+        if matches!(session.transaction(), &TransactionStatus::InTransaction(_))
+            || when == &QueryWhen::Immediately
+        {
+            session.add_transaction_ops(TransactionOps::Peeks(
+                determination.timestamp_context.clone(),
+            ))?;
+        }
+
+        let in_immediate_multi_stmt_txn = session.transaction().is_in_multi_statement_transaction()
+            && when == &QueryWhen::Immediately;
+
+        // If we are in a single statement transaction, there is no need to
+        // acquire read holds to prevent compaction as they will be released
+        // immediately following the completion of the transaction.
+        //
+        // If we're in a multi-statement transaction and the query does not use `AS OF`,
+        // acquire read holds on any sources in the current time-domain if they have not
+        // already been acquired. If the query does use `AS OF`, it is not necessary to
+        // acquire read holds.
+        if in_immediate_multi_stmt_txn {
+            self.check_txn_timedomain_conflicts(
+                session,
+                &determination.timestamp_context,
+                id_bundle,
+            )?;
+
+            let timedomain_id_bundle =
+                self.timedomain_for(source_ids, &timeline_context, session.conn_id(), cluster_id)?;
```

Comment:
This seems like it would fit better on line 2262 right before `let read_holds = ...`. That way we only generate this is we actually use it.